### PR TITLE
add properties for Beyond the Repository project

### PR DESCRIPTION
**Pull request by tdilauro**
This PR updates the BagIt Profiles specification to:
- Add a new optional `description` property for free-form textual description related to Bag-Info tags.
- Add support for specifying lists of allowed payload and tag manifest types/algorithms via the `Manifests-Allowed` and `Tag-Manifests-Allowed` properties, respectively.
- Bump version number.

### CODE REVIEW

**Comment by tdilauro** [comment]
@diamondap, @bbranan,

I haven't received any feedback on this and need to write up the official PR to the [BagIt-Profiles repo](https://github.com/bagit-profiles/bagit-profiles), so I'm adding you as reviewers.

Most of this is, I think, pretty uncontroversial. However, I would especially appreciate your thoughts on whether \*Manifests-Allowed must include any entries in \*Manifests-Required. I'm having second thoughts about it. I included these because of this statement in the current version of the spec: "At least all the tag files listed in Tag-Files-Required must be in included in Tag-Files-Allowed." These are the relevant statements in the updated spec/readme:

- Every algorithm in `Manifests-Required` must also be present in `Manifests-Allowed`, unless `Manifests-Allowed` is specified as `["*"]` or is allowed to default to `["*"]` by omission.
- Every algorithm in `Tag-Manifests-Required` must also be present in `Tag-Manifests-Allowed`, unless `Tag-Manifests-Allowed` is specified as `["*"]` or is allowed to default to `["*"]` by omission.

**Comment by diamondap** [comment]
I agree with the two bullet points. Whatever is in -Required must also be in -Allowed. I'm not sure that * makes sense in Manifests-Allowed / Tag-Manifests-Allowed since those lists are specifically for algorithm names, and only a handful of algorithms have broad support. I think anything in -Required should be explicitly listed in -Allowed.

**Comment by bbranan** [comment]
I agree that anything in `Manifests-Required` must be in `Manifests-Allowed` (and the same for `Tag-Manifests-Required`/`Tag-Manifests-Allowed`).

In the current spec all manifests are allowed, but only some are required (in `Manifests-Required`), so to be backwards compatible with the current spec, we would need to assume that any bag without `Manifests-Allowed` allows all manifests. Allowing a manifest just means it's ok if that manifest is present in the bag; it may or may not be used/processed by whoever is receiving the bag. I believe our goal with `Manifests-Allowed` is to limit the set of possible manifests to those we can be confident can be processed by receiving entities.

I'm not sure that we want to use the `"*"` notation with `Manifests-Allowed`. That notation makes sense for `Tag-Files-Allowed` because there is a need to specify file paths. With `Manifests-Allowed` the only time `"*"` would be used, when everything is allowed, has the same meaning as not including the tag at all. We probably don't want to see values like `sha-*`, but we may be opening that door by using `"*"`. As an alternative, we could just say `If not specified all manifest types are allowed.`

**Comment by tdilauro** [comment]
> I'm not sure that * makes sense in Manifests-Allowed / Tag-Manifests-Allowed since those lists are specifically for algorithm names, and only a handful of algorithms have broad support.

I wasn't particularly concerned about which algorithms should be included in the default case. The current 1.2.0 Bagit Profiles spec doesn't require any particular algorithm and falls back to BagIt spec (any, but at least one) when `Manifests-Required` and/or `Tag-Manifests-Required` are unspecified. I figured the default case in the absence of `Manifests-Allowed` and `Tag-Manifests-Allowed` should be the same unrestricted behavior. Doing other would break backward compatibility.

Even though the BTR profile would not use this behavior, I think the BagIt Profiles spec needs to continue to support it.

As for how to explicitly specify that default behavior, I'm perfectly happy to use an empty list, rather than a list with "*" as its only element.

**Comment by tdilauro** [comment]
Another question: Should `Manifiests-Required` and `Manifests-Allowed` be mutually exclusive? Same for `Tag-Manifiests-Required` and `Tag-Manifests-Allowed`. They don't really make sense together.

We could specify that if the `*Manifests-Required` is present, the  corresponding`*Manifests-Allowed` property should be omitted, but if present will be ignored.

**Comment by bbranan** [comment]
I don't see them as mutually exclusive. The example I had in mind was:
- `Manifests-Required` : "sha-256"
- `Manifests-Allowed` : "md5, sha-256, sha-512"

In this case, you *must* include the sha-256 manifest, but if you want to include one of the other two, that's great, and the ingesting system will know what to do with them.

**Comment by tdilauro** [comment]
Yes, of course you're right, @bbranan. Thanks!

**Comment by diamondap** [comment]
I think it's useful to have both. At APTrust, we require md5 and allow several others. We often get bags with md5 + one other allowed algorithm from the list, probably because the depositor uses that algorithm internally for other processes.

**Comment by tdilauro** [comment]
Thanks, @diamondap.

**Comment by tdilauro** [comment]
@bbranan, @diamondap,

I have updated the PR based on your feedback. Please review and approve if you find the changes acceptable.

Below is the language I plan to eventually use when I submit the PR to Bagit-Profiles repo:

The BagIt Team of the [Beyond the Repository](https://northwestern.app.box.com/s/3qu2qbkdx3aod6wj9jt6977p4byhpj3y) project is recommending changes to the BagIt Profiles specification.

This PR updates the BagIt Profiles specification to:
- Add a new optional `description` property for free-form textual description related to `Bag-Info` tags.
- Add support for specifying lists of allowed payload and tag manifest types/algorithms via the `Manifests-Allowed` and `Tag-Manifests-Allowed` properties, respectively.
- Bump version number.

These changes are non-breaking for v1.2.0, thus packages built to that specification would not require changes (aside from `BagIt-Profile-Identifier` tag in `bag-info.txt`).
- The `description` property of the `Bag-Info` object is optional.
- The `Manifests-Allowed` and `Tag-Manifests-Allowed` lists are optional and reasonable behaviors are defined for the default values.

### Description Property
The `description` property enables the conveyance of any information that might be helpful to either the producer or the consumer of a bag to understand how to provide or interpret the information associated with a particular tag. There is no restriction on the length of `description`, so it may be less than ideal to use as a tooltip for its associated tag.

### Manifests-Allowed and Tag-Manifests-Allowed Properties
With `Manifests-Required` and `Tag-Manifests-Required`, the current BagIt Profiles specification (v1.2.0) supports the enumeration of any algorithms for which a corresponding payload or tag manifest, respectively, MUST be present in a bag. If these lists are specified, though, **all** listed algorithms must be represented.

We needed an additional mechanism through which to designate the set of algorithms supported (rather than required) by the consumer of a bag, so that producers are able create bags appropriately. `Manifests-Allowed` and `Tag-Manifests-Allowed` fulfill this need. If unspecified, these properties impose no constraints on the algorithms available to a producer. If `Manifests-Required` is specified, then if `Manifests-Allowed` is specified, it must include at least those algorithms listed in `Manifests-Required`. The same relationship applies for `Tag-Manifests-Required`/`Tag-Manifests-Allowed`.

**Comment by diamondap** [comment]
Looks good. Thanks for writing that up.

**Comment by bbranan** [comment]
Both the updates to the spec and the proposed PR description of changes look good to me. 👍

**Review by bbranan — APPROVED** [review]

**Comment by tdilauro** [comment]
@bbranan, @diamondap: Would one of you merge this after @diamondap approves his review?

Once that's done, I'll submit the PR to the BagIt-Profiles repo.

Thanks!

**Review by diamondap — APPROVED** [review]

**Comment by bbranan** [comment]
@tdilauro Looking at the final version of this (post merge) it's worth noting that the formatting for the Manifests-Allowed and Tag-Manifests-Allowed sections doesn't look quite right. In both cases the text is contained in a box with a scrollbar. Using the github editor, it seems that this may be due to a leading tab (or set of spaces) at the beginning of each line.

**Comment by tdilauro** [comment]
Thanks, @bbranan. It was an interaction between the unindented version label and the indented text below it. It looks like they may have run into this problem when they upgraded to 1.2.0.

**Pull request merged**
